packetio/buffer.go

```diff
@@ -2,88 +2,171 @@
 package packetio
 
 import (
+	"errors"
 	"io"
 	"sync"
 	"time"
 
 	"github.com/pion/transport/deadline"
 )
 
+var errPacketTooBig = errors.New("packet too big")
+
 // Buffer allows writing packets to an intermediate buffer, which can then be read form.
 // This is verify similar to bytes.Buffer but avoids combining multiple writes into a single read.
 type Buffer struct {
-	mutex   sync.Mutex
-	packets [][]byte
+	mutex sync.Mutex
+
+	// this is a circular buffer.  If head <= tail, then the useful
+	// data is in the interval [head, tail[.  If tail < head, then
+	// the useful data is the union of [head, len[ and [0, tail[.
+	// In order to avoid ambiguity when head = tail, we always leave
+	// an unused byte in the buffer.
+	data       []byte
+	head, tail int
 
 	notify chan struct{}
 	subs   bool
 	closed bool
 
-	// The number of buffered packets in bytes.
-	size int
+	count                 int
+	limitCount, limitSize int
 
-	// The limit on Write in packet count and total size.
-	limitCount int
-	limitSize  int
-
-	// Read deadline timer.
 	readDeadline *deadline.Deadline
 }
 
-// NewBuffer creates a new Buffer object.
+const minSize = 2048
+const cutoffSize = 128 * 1024
+const maxSize = 4 * 1024 * 1024
+
+// NewBuffer creates a new Buffer.
 func NewBuffer() *Buffer {
 	return &Buffer{
 		notify:       make(chan struct{}),
 		readDeadline: deadline.New(),
 	}
 }
 
+// available returns true if the buffer is large enough to fit a packet
+// of the given size, taking overhead into account.
+func (b *Buffer) available(size int) bool {
+	available := b.head - b.tail
+	if available <= 0 {
+		available += len(b.data)
+	}
+	// we interpret head=tail as empty, so always keep a byte free
+	if size+2+1 > available {
+		return false
+	}
+
+	return true
+}
+
+// grow increases the size of the buffer.  If it returns nil, then the
+// buffer has been grown.  It returns ErrFull if hits a limit.
+func (b *Buffer) grow() error {
+	var newsize int
+	if len(b.data) < cutoffSize {
+		newsize = 2 * len(b.data)
+	} else {
+		newsize = 5 * len(b.data) / 4
+	}
+	if newsize < minSize {
+		newsize = minSize
+	}
+	if newsize > maxSize {
```

Comment:
```suggestion
	if b.limitSize <= 0 && newsize > maxSize {
```

Reply:
If that's the semantics that we want, then we can simplify the code by discarding the `maxSize` constant and initialise `limitSize` to 4MB.  But I prefer `maxSize` to be a hard limit.

Are there any legitimate applications to a buffer larger than 4MB?

Reply:
I'm currently don't have actual application that exceeds 4MB limit.
But the size of the streaming contents today is increasing, this change prohibit to the package to be use in such applications.
I can't understand it should be hard limited to 4MB, user should be able to control that I believe.

Reply:
@jech @Sean-Der I'm sorry for delayed reply. But I think this hard limit potentially break compatibility and limit future possibility of the package.